diff --git a/amd/hipcc/src/hipBin_amd.h b/amd/hipcc/src/hipBin_amd.h
index ecea39e07..9cc334a09 100644
--- a/amd/hipcc/src/hipBin_amd.h
+++ b/amd/hipcc/src/hipBin_amd.h
@@ -208,7 +208,7 @@ void HipBinAmd::constructCompilerPath() {
     } else {
       compilerPath = getRoccmPath();
       hipClangPath = compilerPath;
-      hipClangPath /= "lib/llvm/bin";
+      hipClangPath /= "bin";
     }
 
     compilerPath = hipClangPath.string();
